src/llmcompressor/transformers/tracing/GUIDE.md

```diff
@@ -0,0 +1,430 @@
+
+Through reading this guide, you will learn
+1. Why tracing is required when compressing with recipes involving the
+[Sequential Pipeline](/src/llmcompressor/pipelines/sequential/pipeline.py) and modifiers
+such as [GPTQModifier](/src/llmcompressor/modifiers/quantization/gptq/base.py)
+2. How to determine if your model is traceable for your dataset
+3. How to modify your model definition to be traceable
+
+## 1. Why is Tracing Required? ##
+Due to the memory-intensive nature of some modifiers such as [GPTQModifier](/src/llmcompressor/modifiers/quantization/gptq/base.py)
+and [SparseGPTModifier](/src/llmcompressor/modifiers/obcq/base.py), a [Sequential Pipeline](/src/llmcompressor/pipelines/sequential/pipeline.py)
+is required in order to offload activations and reduce memory usage as well as propagate
+the activation error induced by compression.
+
+For example, let's say we want to quantize a basic `3` layer model using the
+[GPTQModifier](/src/llmcompressor/modifiers/quantization/gptq/base.py) and `512`
+calibration samples. The [Sequential Pipeline](/src/llmcompressor/pipelines/sequential/pipeline.py)
+first identifies each of the layers (sequential targets) within the model. Then, the
+pipeline runs each of the `512` samples, one sample at a time, through the first layer.
+When one sample completes its forward pass through the layer, its activations are
+recorded by the [GPTQModifier](/src/llmcompressor/modifiers/quantization/gptq/base.py)
+hessian and the layer output is offloaded to the cpu. After all `512` samples have been
+passed through the layer, the [GPTQModifier](/src/llmcompressor/modifiers/quantization/gptq/base.py)
+uses the recorded activations to compress the weights of the modules within the layer.
+Once module compression is complete, the offloaded activations are used to perform the
+forward pass of the next layer.
+
+The [Sequential Pipeline](/src/llmcompressor/pipelines/sequential/pipeline.py) relies on
+the ability to know what the inputs and outputs are of each model layer in order to
+faithfully "replay" the model's forward pass one layer at a time. While the basic
+assumption that "the input to the next layer is the output of the previous layer" holds
+for most decoder-only LLMs, this assumption breaks for more complex models such as
+encoder-decoder architectures and models which pass RoPE embeddings as an input to each
+layer. For this reason, in order to faithfully capture the data flow between model
+layers, the model's call graph must be traced.
+
+Tracing is done by sampling one batch from the provided dataset, capturing all
+of the operations that are performed during the model forward pass, and encoding those
+operations into a model graph which is then broken into sequential layers to be executed
+independently at calibration time. For a visual example of a model call graph, see
+[Llama_3.2-Vision.svg](/src/llmcompressor/transformers/tracing/assets/Llama_3.2-Vision.svg).
+
+<p align="center">
```

Comment:
What am I supposed to be taking away from this image?

Reply:
TODO for me, this is not a call graph

Reply:
This image depicts the model graph referenced in the above paragraph. The image is a concrete example of what a model graph looks like and helps illustrated what the nodes and edges are within the graph.

Reply:
Yeah I think a small caption/description when showing an image would be helpful to assist the user in what they should be focusing on